### PR TITLE
fix(Core/Chat): Prevent Horde / Alliance chat via custom emotes

### LINKED ISSUE #2099: Horde and Alliance can talk to each other using /e command [$25]
Horde and Alliance can talk to each other using /e command

##### SMALL DESCRIPTION:

Horde and Alliance can talk to each other using /e command

##### EXPECTED BLIZZLIKE BEHAVIOUR:

You shouldn't be able to understand each other, even using /e command

##### CURRENT BEHAVIOUR:

duel
They had a duel
after duel Assok and Seran could completely understand each other using
/e gg hahah (good game)
/e wg sami (going to wintergrasp)
/e wp (well played)
/e sige2 (ok)

##### STEPS TO REPRODUCE THE PROBLEM:

1. Talk to any opposite faction
2. use /e messagehere

##### BRANCH(ES):

master

##### AC HASH/COMMIT:

5aa9266

##### OPERATING SYSTEM:

Ubuntu 16/18

##### OTHER CUSTOMIZATIONS:

<bountysource-plugin>

---
There is a **[$25 open bounty](https://www.bountysource.com/issues/76988714-horde-and-alliance-can-talk-to-each-other-using-e-command?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Stoabrogga**
##### CHANGES PROPOSED:
This is based on the orphaned PR #2117, but I followed a different approach and extended the function "TextEmote" in order to check the faction of all nearby players. If the faction does not match with the player sending the emote only the message "```<player> makes some strange gestures.```" is shown.
Also a new worldserver parameter "AllowTwoSide.Interaction.Emote" has been added which enables / disables emote messages between factions (default is disabled).

##### ISSUES ADDRESSED:
Closes #2099 

##### TESTS PERFORMED:
- tested build on Ubuntu 16.04 / clang 7
- tested successfully in-game

##### HOW TO TEST THE CHANGES:
- try to use normal chat:
  - Horde / Alliance: not readable
  - same faction: readable
- try to use a custom emote (e.g. ```/e hello```)
  - Horde / Alliance:
    - not readable (```AllowTwoSide.Interaction.Emote = 0```)
    - readable (```AllowTwoSide.Interaction.Emote = 1```)
  - same faction: readable

##### KNOWN ISSUES AND TODO LIST:
none

##### Target branch(es):
- [x] Master
 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by pklloveyou** [comment]
Tested and working as expected

**Comment by Stoabrogga** [comment]
I added a worldserver parameter "AllowTwoSide.Interaction.Emote" which allows to enable or disable emote messages between factions (default is disabled). As @FrancescoBorzi correctly stated in the following comment some servers (especially role-play servers) could find it useful to be enabled:
https://github.com/azerothcore/azerothcore-wotlk/issues/2099#issuecomment-513493042

This gives server admins the ability to choose how these emote messages are handled on their server instead of just taking away the option.

@pklloveyou 
If you have the time, could you please take a look at this once more?

**Comment by Stoabrogga** [comment]
Ok, there's still something missing which the original PR also did not address: If the other faction is not allowed to read the emote message there should be another emote text: ```<player> makes some strange gestures.```

I'll try to implement this.

**Comment by pklloveyou** [comment]
c3fbeab  Tested and working as expected

File: src/server/game/Miscellaneous/SharedDefines.h [hunk 1]
```diff
@@ -3559,4 +3559,10 @@ enum NavTerrain
     // we only have 8 bits
 };
 
+enum emotestate
```
Comment (Winfidonarleyan) [inline src/server/game/Miscellaneous/SharedDefines.h]:
```cpp
enum EmoteState
```
Maybe?

**Comment by Stoabrogga** [comment]
@Winfidonarleyan 
I'm now trying a different approach, could you perhaps take a look at the code? I tested this and it seems to work quite well, now correctly showing "```<player> makes some strange gestures.```" when the emote should not be shown to the opposite faction. Btw, this is now also true for such items like the Titanium Seal of Dalaran, which should be correct according to this post:
https://us.battle.net/forums/en/wow/topic/8704130878

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -20945,9 +20945,30 @@ void Player::TextEmote(const std::string& text)
...
+
     WorldPacket data;
-    ChatHandler::BuildChatPacket(data, CHAT_MSG_EMOTE, LANG_UNIVERSAL, this, this, _text);
-    SendMessageToSetInRange(&data, sWorld->getFloatConfig(CONFIG_LISTEN_RANGE_TEXTEMOTE), true, !sWorld->getBoolConfig(CONFIG_ALLOW_TWO_SIDE_INTERACTION_CHAT));
+    std::list<Player*> players;
+    Trinity::AnyPlayerInObjectRangeCheck checker(this, sWorld->getFloatConfig(CONFIG_LISTEN_RANGE_TEXTEMOTE));
+    Trinity::PlayerListSearcher<Trinity::AnyPlayerInObjectRangeCheck> searcher(this, players, checker);
+    this->VisitNearbyWorldObject(sWorld->getFloatConfig(CONFIG_LISTEN_RANGE_TEXTEMOTE), searcher);
+
+    for (std::list<Player*>::const_iterator itr = players.begin(); itr != players.end(); ++itr)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Player/Player.cpp]: Why don't you use?
```cpp
for (auto const& itr : players)
```

Comment (Stoabrogga) [reply src/server/game/Entities/Player/Player.cpp]: Because I'm a C++ noob 🤣 

Thanks, will push a new commit.

**Comment by Stoabrogga** [comment]
@pklloveyou 
Could you please test this PR one last time? It would be especially interesting if the message "```<player> makes some strange gestures.```" is shown in the correct language.

**Comment by pklloveyou** [comment]
Tested and working as expected

**Comment by Stoabrogga** [comment]
@pklloveyou 
Thanks! Was the translation of the message ok?

**Comment by pklloveyou** [comment]
Simplified Chinese normal display

**Comment by Stoabrogga** [comment]
Ok, so I think this is ready to be merged. Thank you @pklloveyou @Winfidonarleyan

**Pull request merged**